### PR TITLE
#1142  - Create ES latest build

### PR DESCRIPTION
**Pull request by treshugart**
This creates an ES2015 build in the `es/` folder, as opposed to using NWB to do ES5 + ESM. The downside is that consumers must still transpile the source from ES2015 to their target. The upside is it reduces complexity of having `es/` and `es-latest/`.

Maybe we should consider publishing both? The main question here is how would one consume `module` from the `package.json` and transpile it with little friction. It seems they'd have to include `node_modueles/skatejs` in their babel config? This is why I liked the ES5 + ESM bundle; it worked well with the `module` field. Either way the consumer must include `node_modules/skatejs`, I guess.

Thoughts welcome.

### CODE REVIEW

**Comment by matthewp** [comment]
Does this add `.js` to module specifiers?

**Comment by treshugart** [comment]
@matthewp not yet. That still needs to be done.

**Comment by matthewp** [comment]
Sure, fine by me. Just making sure that the plan is for `es/` to be for web usage as well.  Do we just need to figure out that problem you discovered with babel-plugin-modules-web-compat, right?

**Comment by treshugart** [comment]
> Sure, fine by me. Just making sure that the plan is for es/ to be for web usage as well. Do we just need to figure out that problem you discovered with babel-plugin-modules-web-compat, right?

I'd like to make it consumable by both web and bundlers. Using suffixes shouldn't be an issue, but we'll need to make an opinion about peer deps like Preact and where we'll look them up. As discussed in https://github.com/matthewp/babel-plugin-modules-web-compat/pull/1 it seems as though we should assume that they'd be siblings. However that doesn't work if you have an npm org, so may have to do some org detection on both the core package and its deps.

File: nwb.config.js [hunk 1]
```diff
@@ -1,19 +1,53 @@
...
+const path = require('path');
+const rimraf = require('rimraf');
+const yargs = require('yargs');
+
+const browsers = require('./test/browsers');
+
+const [ cmd ] = yargs.argv._;
+const isBuild = cmd === 'build';
+const babelConfig = {
+  plugins: ['transform-react-jsx']
```
Comment (Hotell) [inline nwb.config.js]: hmm why is this needed? we aren't using jsx in source code are we ?

File: nwb.config.js [hunk 2]
```diff
@@ -1,19 +1,53 @@
...
+const fs = require('fs');
+const glob = require('glob');
+const mkdirp = require('mkdirp');
+const path = require('path');
+const rimraf = require('rimraf');
+const yargs = require('yargs');
+
+const browsers = require('./test/browsers');
+
+const [ cmd ] = yargs.argv._;
```
Comment (Hotell) [inline nwb.config.js]: huh, isn't there any sugar feature passing args like in pure webpack ?

File: nwb.config.js [hunk 3]
```diff
@@ -1,19 +1,53 @@
...
+const isBuild = cmd === 'build';
+const babelConfig = {
+  plugins: ['transform-react-jsx']
+};
+
+if (isBuild) {
+  const files = glob.sync('src/**/*.js');
+  rimraf.sync('./es');
+  files.forEach(f => {
```
Comment (Hotell) [inline nwb.config.js]: isn't there some easier way how to do this? aren't there some babel-cli flags for that? sorry for dumb questions, I'm just not used to for such kind of complexity to just emit raw es6 to some other folder (  in TS this is just one liner from cli :-| )

File: nwb.config.js [hunk 4]
```diff
@@ -1,19 +1,53 @@
...
+    const targetDir = path.dirname(target);
+    mkdirp.sync(targetDir);
+    fs.writeFileSync(target, code);
+  });
+}
 
 module.exports = {
   type: 'web-module',
   npm: {
+    // CJS is kinda useless if we have UMD.
```
Comment (Hotell) [inline nwb.config.js]: 👍

File: nwb.config.js [hunk 5]
```diff
@@ -1,19 +1,53 @@
...
     cjs: false,
-    esModules: true,
+
+    // We create a custom ESBuild above.
```
Comment (Hotell) [inline nwb.config.js]: I'm confused, so the custom ESBuild will produce  es6 modules with transiled code to es5 ?
-> ` presets: ['stage-0', 'es2017', 'es2016']`

If so, this doesn't make any sense -> we don't want classes to be transformed to function

Comment (treshugart) [inline nwb.config.js]: Each stage converts it down to the syntax of the latest spec. Each spec preset converts it to the previous year. So ES2016 compiles down to ES2015.

File: nwb.config.js (hunk 2 above)
Comment (treshugart) [inline nwb.config.js]: https://github.com/yargs/yargs is sugar

File: nwb.config.js (hunk 1 above)
Comment (treshugart) [inline nwb.config.js]: We use it in tests, but don't need it for prod builds. The tests and bundles share the NWB babel config,

File: nwb.config.js (hunk 3 above)
Comment (treshugart) [inline nwb.config.js]: > I'm just not used to for such kind of complexity to just emit raw es6 to some other folder ( in TS this is just one liner from cli :-| )

Let's set the TypeScript stuff aside, please. I think I can probably work out a simple command for it, I was just sketching off the top of my head.

Comment (treshugart) [inline nwb.config.js]: Command found and implemented.

File: package.json [hunk 6]
```diff
@@ -9,7 +9,6 @@
     "url": "https://github.com/skatejs/skatejs"
   },
   "main": "umd/skatejs.js",
-  "module": "es/index.js",
```
Comment (Hotell) [inline package.json]: so with this I am unable to tree shake skate right? I would have to explicitly alias it in webpack for instance ( with which I'd have to use native-shim though )

File: package.json [hunk 7]
```diff
@@ -53,7 +56,7 @@
   "scripts": {
     "docs:build": "gitbook install && gitbook build",
     "docs:watch": "gitbook install && gitbook serve",
-    "prepublish": "nwb build",
+    "prepublish": "nwb build && babel src --out-dir es --presets=stage-0,es2017,es2016",
```
Comment (Hotell) [inline package.json]: 👍
